edi_shipment.py

```diff
@@ -0,0 +1,615 @@
+            else:
+                getattr(shipment_edi, 'read_%s' % msg_id)(line)
+
+        # invoice_line.search_related(shipment)
+        return shipment_edi
+
+    def add_attachment(self, attachment, filename=None):
+        pool = Pool()
+        Attachment = pool.get('ir.attachment')
+        if not filename:
+            filename = datetime.now().strftime("%y/%m/%d %H:%M:%S")
+        attach = Attachment(
+            name=filename,
+            type='data',
+            data=attachment.decode('utf8'),
+            resource=self)
+        attach.save()
+
+    @classmethod
+    def import_shipment_in(cls, edi_shipments=None):
+        pool = Pool()
+        Configuration = pool.get('stock.configuration')
+        configuration = Configuration(1)
+        source_path = os.path.abspath(configuration.inbox_path_edi or
+             DEFAULT_FILES_LOCATION)
+
+        files = [os.path.join(source_path, fp) for fp in
+                 os.listdir(source_path) if os.path.isfile(os.path.join(
+                     source_path, fp))]
+        files_to_delete = []
+        to_save = []
+        attachments = dict()
+        for fname in files:
+            if fname[-4:].lower() not in KNOWN_EXTENSIONS:
+                continue
+            with open(fname, 'r', encoding='latin-1') as fp:
+                data = fp.readlines()
+                shipment = cls.import_edi_file([], data)
+
+            basename = os.path.basename(fname)
+            if shipment:
+                attachments[shipment] = ("\n".join(data), basename)
+                to_save.append(shipment)
+                files_to_delete.append(fname)
+
+        if to_save:
+            cls.save(to_save)
+
+        # with Transaction().set_user(0, set_context=True):
+        #     for shipment, (data, basename) in attachments.items():
+        #         shipment.add_attachment(data, basename)
+
+        if files_to_delete:
+            for file in files_to_delete:
+                os.remove(file)
+
+        cls.search_references(to_save)
+
+    def _get_new_lot(self, line, quantity):
+        pool = Pool()
+        Lot = pool.get('stock.lot')
+
+        if line.expiration_date:
+            today = datetime.today().date()
+            lot = Lot()
+            lot.product = line.product
+            lot.expiration_date = line.expiration_date
+            lot.on_change_product()
+            return lot
+
+    @classmethod
+    @ModelView.button
+    def search_references(cls, edi_shipments):
+        Line = Pool().get('edi.shipment.in.line')
+        to_save = []
+        for edi_shipment in edi_shipments:
+            if edi_shipment.shipment:
+                continue
+            for eline in edi_shipment.lines:
+                eline.search_related(edi_shipment)
+                to_save.append(eline)
+        Line.save(to_save)
+
+    @classmethod
+    @ModelView.button
+    def create_shipment(cls, edi_shipments):
+        ShipmentIn = Pool().get('stock.shipment.in')
```

Comment:
Agrupar amb una variable "pool":

    pool = Pool()
    ShipmentIn = pool.get('stock.shipment.in')
    .....